rasa/__init__.py

```diff
@@ -1,10 +1,17 @@
 import logging
 
-from rasa import version  # noqa: F401
+from rasa import version, plugin  # noqa: F401
 from rasa.api import run, train, test  # noqa: F401
 
 # define the version before the other imports since these need it
 __version__ = version.__version__
 
 
 logging.getLogger(__name__).addHandler(logging.NullHandler())
+
+try:
+    import rasa_plus
+
+    rasa_plus.init_hooks()
```

Comment:
OPTIONAL: There is a small improvement that just came to mind which I think might avoid us running into cyclic imports in the future:
* we could move this try block into the `plugin.py` module
* in that module rather than defining the plugin_manager using a variable, we can use a mememoized function, e.g. along the lines of 
```python
__plugin_manager = None

def plugin_manager():
    if __plugin_manager:
        return __plugin_manager

    __plugin_manager = pluggy.PluginManager("rasa")
    __plugin_manager.add_hookspecs(sys.modules["rasa.plugin"])
    _discover_plugins(__plugin_manager)
    return __plugin_manager

def _discover_plugins(manager):
    try:
        import rasa_plus
        rasa_plus.init_hooks(manager)
    except ModuleNotFoundError:
        pass
```
this should avoid cyclic imports as the initialization isn't done during import but rather on the first usage of the plugin manager. 

the `_discover_plugins` function could also use the setup tools feature for plugins: https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins
using `rasa_plugin_eps = entry_points(group='rasa.plugin_init')` to make the registration more generic. If you want you can add that (it avoids explicitly mentioning rasa_plus) but I don't think it is required and a nice to have.

Reply:
I'll do the refactoring of the first part and read up on setuptools entry points, at a quick glance I'm not sure what the `plugin_init` refers to in your example 🤔 Shouldn't it be _discover_plugins?

Reply:
Silly question: why do we need the `__plugin_manager` global variable and then that if check  in plugin_manager? Is that to avoid reinitialising the `PluginManager` and doing the hook registration etc over and over again every time a rasa command is run?

Reply:
You mean the `rasa.plugin_init` that would go into the `pyprojects.toml` in this line `rasa_plugin_eps = entry_points(group='rasa.plugin_init')`?

This is just a label, it allows to then "search" for all the plugins that have defined this "group" in their pyproject.toml.

E.g. `rasa_plus` would need to add something like this to the pyproject.toml:
```toml
# Note the quotes around timmins.display in order to escape the dot 
[project.entry-points."rasa.plugin_init"]
excl = "rasa_plus:init_hooks"
```

and in rasa rather than doing the 
```python
def _discover_plugins(manager):
    try:
        import rasa_plus
        rasa_plus.init_hooks(manager)
    except ModuleNotFoundError:
        pass
```
we'd use the group to discover the plugin:
```python
def _discover_plugins(manager):
    rasa_plugin_eps = entry_points(group='rasa.plugin_init')
    for plugin in rasa_plugin_eps:
        plugin_init = plugin.load()
        plugin_init(manager)
```

Reply:
> Silly question: why do we need the __plugin_manager global variable and then that if check in plugin_manager? Is that to avoid reinitialising the PluginManager and doing the hook registration etc over and over again every time a rasa command is run?

not a silly question at all and you hit the nail on the head. it avoids to recreate the plugin manager multiple times. now that i think about it, it might even work without the memoization but i am not sure if we should rely on that.

IMPORTANT: I just noticed that i missed to add the `global` in the example above:
```
def plugin_manager():
    global __plugin_manager

    if __plugin_manager:
        return __plugin_manager

    __plugin_manager = pluggy.PluginManager("rasa")
    __plugin_manager.add_hookspecs(sys.modules["rasa.plugin"])
    _discover_plugins(__plugin_manager)
    return __plugin_manager
```

Reply:
to avoid the need to mess with the global variable, you can also directly use functools:
https://docs.python.org/3/library/functools.html#functools.cache

(probably better anyways)

Reply:
Ahaha, PyCharm complained about the unresolved reference to __plugin_manager, so added the global bit nonetheless. I like the functools.cache approach 😎

Reply:
i am only suggesting untested code written without an editor :trollface:

Reply:
Ah functools.cache is new in 3.9, I'll use lru_cache which is available in all python versions, hope that's fine.
Also, the importlib.metadata.entry_points function doesn't accept group argument as stated in the setuptools docs (at least not in Python versions we're using), so had to amend that a bit.

Reply:
Ah shoot, we can't use the entrypoint set-up because importlib.metadata is new in Python 3.8 🙈